canopen/source/co_sync.c

```diff
@@ -182,31 +188,181 @@ void COSyncRestart(CO_SYNC *sync)
 /*
 * see function definition
 */
-CO_ERR COTypeSyncIdWrite(struct CO_OBJ_T *obj, struct CO_NODE_T *node, void *buf, uint32_t len)
-{
-    CO_ERR    result = CO_ERR_NONE;
-    uint32_t  nid;
-    uint32_t  oid;
+void COSyncProdActivate(CO_SYNC *sync) {
+    uint32_t ticks, time;
+    CO_ERR   err;
+    CO_NODE *node;
+    int16_t  tid;
+
+    node = sync->Node;
+
+    err = CODictRdLong(&node->Dict, CO_DEV(0x1006, 0), &sync->Cycle);
+    if (err != CO_ERR_NONE) {
+        node->Error = CO_ERR_CFG_1006_0;
+        sync->Cycle = 0;
+        return;
+    }
+
+    if ((sync->Cycle == 0) || ((sync->CobId & CO_SYNC_COBID_OFF) == 0)) {
+        /* SYNC producer is disabled */
+        return;
+    }
+
+    time = COTmrGetMinTime(&node->Tmr, CO_TMR_UNIT_1MS);
+    if ((time * 1000) > sync->Cycle) {
+        /* Provided timer driver has small resolution for configured value */
+        node->Error = CO_ERR_SYNC_RES;
+        return;
+    }
 
-    (void)len;
-    (void)node;
-    nid = *(uint32_t*)buf;
-    (void)COObjRdDirect(obj, &oid, CO_LONG);
+    if (sync->Tmr >= 0) {
+        tid = COTmrDelete(&node->Tmr, sync->Tmr);
+        if (tid < 0) {
+            node->Error = CO_ERR_TMR_DELETE;
+        }
+    }
 
-    /* currently generating SYNCs is not supported */
-    if((nid & CO_SYNC_COBID_OFF) != (uint32_t)0) {
-        result = CO_ERR_OBJ_RANGE;
+    time = (sync->Cycle / 1000);
+    if (time > 0) {
+        ticks = COTmrGetTicks(&node->Tmr, time, CO_TMR_UNIT_1MS);
+        sync->Tmr = COTmrCreate(&node->Tmr,
+            ticks,
+            ticks,
+            COSyncProdSend,
+            sync);
+        if (sync->Tmr < 0) {
+            node->Error = CO_ERR_TMR_CREATE;
+        }
+    } else {
+        sync->Tmr = -1;
     }
+}
+
+/*
+* see function definition
+*/
+void COSyncProdDeactivate(CO_SYNC *sync) {
+    int16_t tid;
+
+    if (sync->Tmr >= 0) {
+        tid       = COTmrDelete(&sync->Node->Tmr, sync->Tmr);
+        sync->Tmr = -1;
+        if (tid < 0) {
+            sync->Node->Error = CO_ERR_TMR_DELETE;
+        }
+    }
+}
+
+/*
+* see function definition
+*/
+void COSyncProdSend(void *parg) {
+    CO_IF_FRM  frm;
+    CO_SYNC   *sync;
+
+    sync = (CO_SYNC *) parg;
+
```

Comment:
At this place, I think we should check that SYNC messages are allowed in current operating mode (e.g. skipped in STOP mode). For example:

```c
    CO_NMT *nmt;

    sync = (CO_SYNC *)parg;
    nmt  = sync->Node.Nmt;
    if ((nmt->Allowed & CO_SYNC_ALLOWED) == 0) {
        return;
    }
```

Reason: with this strategy, we can define the allowed messages for each operating mode at central place in `co_nmt.c`